cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -85,67 +87,114 @@ message Result {
+In order to enforce L7 DNS policy, SDP needs to translate IP address to endpoint ID and identity ID. The simplest option is to reuse the same gRPC stream and implement another method to get the mappings from cilium agent. Since the actual source of truth are the bpf maps. And SDP is expected to resolve the DNS queries even when agent is down, it is better to read the information directly from the bpf maps. We choose the bpf option as it is more reliable and does not rely on agent availability. It also aligns with having a HA DNS proxy.
 
+#### Option 1a: gRPC method
 
-### Option 1: gRPC method
+Get the endpoint ID and identity for a given IP address by making a gRPC call to cilium agent.
 
-#### Pros
+##### Pros
 
 * Avoid interacting with low-level details. Simpler to maintain and stay upto date with any datapath changes.
 * All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
 
-#### Cons
+##### Cons
 
 * Reliance on gRPC call in the hot-path
+* Does not cache the data for endpoints that never had a DNS query.
 * In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
 
-### Option 2: Read from bpf maps
+#### Option 1b: Listen to ipcache updates via grpc
 
-#### Pros
+Similar to envoy, SDP can listen to ipcache updates and maintain a local cache of IP to endpoint ID mappings via grpc. This changes the way DNS proxy currently behaves i.e pull based model to push based model.
+
+##### Pros
+
+* Simpler to maintain and stay upto date with any new ip<>identity changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Need a method for reconciling the cache in case of SDP restarts.
+* SDP will be aware of all the endpoints data, which might not be needed for DNS proxy.
+
+#### Option 2: Read from bpf maps [Recommended]
+
+We will need to read from couple of bpf maps to get the endpoint ID and identity for a given IP address. For ip to endpoint ID, we can read from `cilium_lxc` and for ip to identity, we can read from `cilium_ipcache`.
+
+##### Pros
 
 * Reduced resource utilization in agent since agent doesn't need to process rpc calls in hotpath
 * IP to endpoint ID or identity translation does not rely on agent availability.
+* DNS can be resolved even when agent is down.
 
-#### Cons
+##### Cons
 
 * Low level interactions from components external to cilium agent.
 
+### Q: Why did we choose to read from bpf maps ?
+
+The reason we chose to read from bpf maps is because the DNS proxy should be able to resolve the DNS queries even when agent is down. The bpf maps are the source of truth for cilium agent and are updated in real time. So, reading from bpf maps ensures that the DNS proxy is always in sync with the cilium agent.
+
+### Flag to disable built-in proxy
```

Comment:
Did we check on how hard it is to decouple the endpoint restoration logic from built in proxy ?

Reply:
Added that we try to restore the dns rules from in memory as they have already been parsed from the file system instead of relying on standalone dns proxy.